src/test/java/teammates/test/cases/browsertests/FeedbackRankQuestionUiTest.java

```diff
@@ -102,107 +142,337 @@ public void testStudentSubmitAndResultsPages() throws Exception {
+        assertEquals("All allocated ranks are different.\n"
+                    + "At least 3 recipients have been ranked.\n"
+                    + "At most 3 recipients have been ranked.",
+                    submitPage.getRankMessage(qnNumber, 3));
+
+
         submitPage.selectResponseTextDropdown(qnNumber, 3, 0, "4");
-        assertEquals("Rank no more than 3 recipients.", submitPage.getRankMessage(qnNumber, 3));
+        assertEquals("All allocated ranks are different.\n"
+                    + "At least 3 recipients have been ranked.\n"
+                    + "You can rank at most 3 recipients. You have ranked 1 extra recipients.",
+                    submitPage.getRankMessage(qnNumber, 3));
+
         submitPage.selectResponseTextDropdown(qnNumber, 0, 0, "");
-        assertTrue("No error message expected", submitPage.getRankMessage(qnNumber, 3).isEmpty());
+        assertEquals("All allocated ranks are different.\n"
+                    + "At least 3 recipients have been ranked.\n"
+                    + "At most 3 recipients have been ranked.",
+                    submitPage.getRankMessage(qnNumber, 3));
+
+        // Question with different min and max options to be ranked restriction
+        qnNumber = 15;
+        submitPage.selectResponseTextDropdown(qnNumber, 0, 0, "1");
+        assertEquals("All allocated ranks are different.\n"
+                    + "You need to rank at least 2 recipients. You have ranked 1 recipients so far.\n"
+                    + "At most 4 recipients have been ranked.",
+                    submitPage.getRankMessage(qnNumber, 3));
+
+        submitPage.selectResponseTextDropdown(qnNumber, 1, 0, "2");
+        assertEquals("All allocated ranks are different.\n"
+                    + "At least 2 recipients have been ranked.\n"
+                    + "At most 4 recipients have been ranked.",
+                    submitPage.getRankMessage(qnNumber, 3));
+
+        submitPage.selectResponseTextDropdown(qnNumber, 2, 0, "3");
+        assertEquals("All allocated ranks are different.\n"
+                    + "At least 2 recipients have been ranked.\n"
+                    + "At most 4 recipients have been ranked.",
+                    submitPage.getRankMessage(qnNumber, 3));
+
+
+        submitPage.selectResponseTextDropdown(qnNumber, 3, 0, "3");
+        assertEquals("Multiple recipients are given the same rank! eg. 3.\n"
+                    + "At least 2 recipients have been ranked.\n"
+                    + "At most 4 recipients have been ranked.",
+                    submitPage.getRankMessage(qnNumber, 3));
+
+        submitPage.selectResponseTextDropdown(qnNumber, 3, 0, "4");
+        assertEquals("All allocated ranks are different.\n"
+                    + "At least 2 recipients have been ranked.\n"
+                    + "At most 4 recipients have been ranked.",
+                    submitPage.getRankMessage(qnNumber, 3));
 
```

Comment:
what about ranking extra?

Reply:
@wenmogu sorry completely missed these comments. Can you clarify what you mean by ranking extra ?